Parser/pegen.c

```diff
@@ -2383,28 +2383,81 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
         return _PyAST_Constant(res, NULL, lineno, col_offset, end_lineno, end_col_offset, p->arena);
     }
 
-    asdl_expr_seq* values = _Py_asdl_expr_seq_new(n_elements, p->arena);
-    if (values == NULL) {
+    if (unicode_string_found && !f_string_found && len == 1) {
+        return asdl_seq_GET(strings, 0);
+    }
+
+    asdl_expr_seq* flatenned = _Py_asdl_expr_seq_new(n_flattened_elements, p->arena);
+    if (flatenned == NULL) {
         return NULL;
     }
 
+    /* build flatenned list */
     Py_ssize_t current_pos = 0;
     Py_ssize_t j = 0;
     for (i = 0; i < len; i++) {
         expr_ty elem = asdl_seq_GET(strings, i);
         if (elem->kind == Constant_kind) {
-            asdl_seq_SET(values, current_pos++, elem);
+            asdl_seq_SET(flatenned, current_pos++, elem);
         } else {
             for (j=0; j < asdl_seq_LEN(elem->v.JoinedStr.values); j++) {
                 expr_ty subvalue = asdl_seq_GET(elem->v.JoinedStr.values, j);
                 if (subvalue == NULL) {
                     return NULL;
                 }
-                asdl_seq_SET(values, current_pos++, subvalue);
+                asdl_seq_SET(flatenned, current_pos++, subvalue);
             }
         }
     }
- 
+
+    /* calculate folded element count */
+    Py_ssize_t n_elements = 0;
+    int prev_is_constant = 0;
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flatenned, i);
+        if (!(prev_is_constant && elem->kind == Constant_kind)) {
+            n_elements++;
+        }
+        prev_is_constant = elem->kind == Constant_kind;
+    }
+
+    asdl_expr_seq* values = _Py_asdl_expr_seq_new(n_elements, p->arena);
+    if (values == NULL) {
+        return NULL;
+    }
+
+    /* build folded list */
+    _PyUnicodeWriter writer;
+    current_pos = 0;
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flatenned, i);
+        if (elem->kind == Constant_kind) {
+            _PyUnicodeWriter_Init(&writer);
+            for (j = i; j < n_flattened_elements; j++) {
+                expr_ty elem = asdl_seq_GET(flatenned, j);
+                if (elem->kind == Constant_kind) {
+                    _PyUnicodeWriter_WriteStr(&writer, elem->v.Constant.value);
+                } else {
+                    break;
+                }
+            }
+            i = j-1;
+            expr_ty concat_str = _PyAST_Constant(_PyUnicodeWriter_Finish(&writer),
+                                                 NULL, lineno, col_offset,
+                                                 end_lineno, end_col_offset, p->arena);
```

Comment:
Question: what does these positions belong to? Wouldn't it be more precesine if we `first_elem->lineno, first_elem->col_offset, last_elem->end_lineno, last_elem->end_col_offset`?

Reply:
Can you check if it looks alright now?